src/server.rs

```diff
@@ -308,21 +306,29 @@ impl Default for Server {
     }
 }
 
-async fn merge_headers(
-    global_headers: &Arc<Headers>,
-    request_headers: &HeaderMap,
-) -> HashMap<String, String> {
-    let mut headers = HashMap::new();
+async fn apply_middleware(
+    request: &mut Request,
+    middleware_type: MiddlewareRoute,
+    middleware_router: &MiddlewareRouter,
+    route_uri: &str,
```

Comment:
@AntoineRR , can you please explain this function a bit?

Reply:
Sure!
We have two middlewares to apply, one before routing and another after it.
We first route on the middleware function and execute it, which gives us some modifications to apply to the `request` object. We only care about the headers for now => they become the request headers.
Therefore the middleware could modify the request (at least the "before" middleware for Robyn currently), which is why `request` is `&mut`.
For the "after" middleware, I think it cannot currently modify the request or the response, so the `execute_midlleware_function` should return an empty hashmap and we will just have executed the python function.
The idea here is just to factorize the code to execute both middlewares so it is easier to understand and maintain. Although if you think it makes sense to keep the execution of both middleware separate it is also possible

Reply:
Perfect. @AntoineRR do you think we can modify request body in `before_request` and response body in `after_request`?

Reply:
I think it would make sense yes. Maybe introducing a common trait for `Request` and `Response` could help us with this. The `apply_middleware` method would become generic and take either a `Request` or a `Response`, and the trait will then provide a method for parsing the hashmap returned by `execute_middleware_function` and apply it to the object.

Reply:
I could try to do this in an other PR, as it will add a new feature for Robyn, i.e. response modification in "after" middleware. What do you think of it?

Reply:
> I could try to do this in an other PR, as it will add a new feature for Robyn, i.e. response modification in "after" middleware. What do you think of it?

Sounds great. We implemented some of it here too: https://github.com/sansyrox/robyn/pull/297

Just in case you need any reference.


> Maybe introducing a common trait for Request and Response could help us with this

What kind of trait will it be?

Reply:
Thanks for the reference :)
This could be something like this I guess:
```
trait Data {
     fn modify_from_hashmap(&mut self, modifications: Hashmap<String, String>) -> &mut self;
}
```
and then `apply_middleware` signature would become:
```
async fn apply_middleware<T: Data>(
    data: &mut T,
    middleware_type: MiddlewareRoute,
    middleware_router: &MiddlewareRouter,
    route_uri: &str,
)
```
After writing impl of `Data` for `Request` and `Response`, they could be used as the `data` parameter for the function. Then we call `data.modify_from_hashmap` with the hashmap from `execute_middleware_function` and we're good! Using generics in Rust is a zero cost abstraction due to monomorphization so there will be no performance impact from this. Using this solution makes modifying requests and response from a hashmap easier IMO, and everything is contained inside the `Request` and `Response` struct.
I don't like the name `Data` for the trait but it is just an example here.

Reply:
> Using generics in Rust is a zero cost abstraction due to monomorphization so there will be no performance impact from this.

I wasn't aware of this. Thank you! 😄